src/lib/krb5/ccache/ccselect.c

```diff
@@ -132,14 +132,39 @@ krb5_cc_select(krb5_context context, krb5_principal server,
     struct ccselect_module_handle **hp, *h;
     krb5_ccache cache;
     krb5_principal princ;
+    krb5_principal srvcp = NULL;
+    char **fbrealms = NULL;
 
     *cache_out = NULL;
     *princ_out = NULL;
 
     if (context->ccselect_handles == NULL) {
         ret = load_modules(context);
         if (ret)
-            return ret;
+            goto cleanup;
+    }
+
+    /* Try to use the fallback host realm for the server if there's no
+     * authoritative realm. */
+    if (krb5_is_referral_realm(&server->realm) &&
+        server->type == KRB5_NT_SRV_HST && server->length == 2 &&
+        server->data[1].length != 0) {
+        ret = krb5_copy_principal(context, server, &srvcp);
+        if (ret)
+            goto cleanup;
+
+        server = srvcp;
+        ret = krb5_get_fallback_host_realm(context, &server->data[1],
+                                           &fbrealms);
+        if (ret) {
+            goto cleanup;
+        }
+
+        if (fbrealms != NULL && fbrealms[0] != NULL) {
```

Comment:
Actually, looking at the contract of krb5_get_fallback_host_realm() and how it is used in get_creds.c, I don't think we need this check at all.  If krb5_get_fallback_host_realm() succeeds it will return at least one entry.